include/zephyr/sys/util.h

```diff
@@ -562,6 +562,19 @@ static inline uint8_t bin2bcd(uint8_t bin)
  */
 uint8_t u8_to_dec(char *buf, uint8_t buflen, uint8_t value);
 
+/**
+ * @brief Sign extend an 8, 16 or 32 bit value using the index bit as sign bit.
+ *
+ * @param value The value to sign expand.
+ * @param index 0 based bit index to sign bit (0 to 31)
+ */
+static inline int32_t sign_extend(int32_t value, int index)
```

Comment:
Might be worth considering redefining as a macro with some `__typeof__()` magic to work with arbitrary input word sizes.   Because you just know that someone is going to ask for a sign_extend_64().

Reply:
Yeah I tried something like
```c
#define SIGN_EXTEND(value, index) \
       (__typeof__(value))(value << (NUM_BITS(value) - 1 - index)) >> (NUM_BITS(value) - 1 - index)
```

But then realized that I need the signed version of the type, not the same type, for that to work. :-) If you have any idea... meanwhile I've added a 64 bit variants to anticipate that, Linux has it so why not.